### ISSUE TITLE
Functions to create callbacks

### ISSUE THREAD

**Issue by Alphish** [Alphish/gm-community-toolbox#25]
Basically, I want to introduce functions that generate callbacks with given functions, for later use or to be passed into functions that operate on other functions.

Functions I came up with:
- `callback`, with the first argument being the script/method, and the subsequent arguments being the parameters to call it with
- `callback_ext`, with the first argument being the script/method, and the second argument being an array of arguments
- `callback_item` to make a parameterised callback, with the first argument being the script/method, and the subsequent arguments being the additional parameters beside the item
- `callback_item_ext` to make a parameterised callback, with the first argument being the script/method, the second argument being the non-item arguments, and the third argument being the index of the item in the function (0 by default)

Usage examples:
```gml
button.title = "New game";
button.on_click = callback(room_goto, rm_Level1);
```
```gml
function filter_by_grade(_some_array, _min_grade, _max_grade) {
    // is_between has a form of "is_between(value,min,max)"
    return array_filter(_some_array, callback_item(is_between, _min_grade, _max_grade));
}
```

I didn't think too hard about the function names, so suggestions are welcome here. ^^

**Comment by Mercerenies** [comment]
I don't think this can be done until we get real closures in GM, but as we've discussed it sounds like that's coming soon-ish.

The mathematical term for this is [*partial application*](https://en.wikipedia.org/wiki/Partial_application). The term ["currying"](https://en.wikipedia.org/wiki/Currying) is also frequently understood in this context, even though that's not exactly what's happening here.

C++ calls your `callback` function [`bind`](https://en.cppreference.com/w/cpp/utility/functional/bind) and doesn't have any `callback_ext` (being statically typed and all). Python uses [`partial`](https://docs.python.org/3/library/functools.html#functools.partial) for the same.

Personally, I'm fond of `curry`, since it's pretty impossible to mistake that function name for anything else (unless you happen to be writing a game about Indian-style cooking). So we could go with `curry` and `curry_ext`. Then the forms that leave one argument out could be `curry1`, `curry1_ext`. It's probably worth providing a two-argument bind as well, since built-ins like `array_sort` expect a two-argument callable. Going three or bigger is likely excessive though.

The name "curry", like our `eucmod` discussion, has the benefit of being immediately recognizable to anyone who knows what the word means while not being mistaken for something else by those who don't. "bind" and "partial" both have lots of other meanings, and "callback", while it describes the intended use case well, doesn't immediately imply to me what the function in question is doing.
